flow/project.py

```diff
@@ -326,20 +345,98 @@ def cmd(self):
 class JobOperation(_JobOperation):
-    pass
+    """This class represents the information needed to execute one group for one job.
+
+    The execution or submission of a :py:class:`FlowGroup` uses a passed in command
+    which can either be a string or function with no arguments that returns a shell
+    executable command.  The shell executable command won't be used if it is
+    determined that the group can be executed without forking.
+
+    .. note::
+
+        This class is used by the :class:`~.FlowGroup` class for the execution and
+        submission process and should not be instantiated by users themselves.
+
+    :param id:
+        The id of this JobOperation instance. The id should be unique.
+    :type id:
+        str
+    :param name:
+        The name of the JobOperation.
+    :type name:
+        str
+    :param job:
+        The job associated with this operation.
+    :type job:
+        :class:`signac.contrib.job.Job`
+    :param cmd:
+        The command that executes this operation. Can be a callable that when
+        evaluated returns a string.
+    :type cmd:
+        callable or str
+    :param directives:
+        A dictionary of additional parameters that provide instructions on how
+        to execute this operation, e.g., specifically required resources.
+    :type directives:
+        :class:`dict`
+    """
+    def __init__(self, id, name, job, cmd, directives=None):
+        self._id = id
+        self.name = name
+
+        assert isinstance(job, signac.contrib.job.Job)
+
+        self._jobs = [job]
+        if not (callable(cmd) or isinstance(cmd, str)):
+            raise ValueError("JobOperation cmd must be a callable or string.")
+        self._cmd = cmd
+
+        if directives is None:
+            directives = dict()  # default argument
+        else:
+            directives = dict(directives)  # explicit copy
+
+        # Keys which were explicitly set by the user, but are not evaluated by the
+        # template engine are cause for concern and might hint at a bug in the template
+        # script or ill-defined directives. We are therefore keeping track of all
+        # keys set by the user and check whether they have been evaluated by the template
+        # script engine later.
+        keys_set_by_user = set(directives)
+
+        # We use a special dictionary that allows us to track all keys that have been
+        # evaluated by the template engine and compare them to those explicitly set
+        # by the user. See also comment above.
+        self.directives = TrackGetItemDict(
+            {key: value for key, value in directives.items()})
+        self.directives._keys_set_by_user = keys_set_by_user
+
+    @property
+    def job(self):
+        # The length of `_jobs` will never be more than one due to the
+        # instance check we're doing while initializing an instance of
+        # JobOperation. Hence return only the first element of `self._jobs`
+        return self._jobs[0]
```

Comment:
Both the function itself as well as the comment appear outdated?

Reply:
This comment here is not outdated.
We deprecated `JobOperation` in #325 and we copied the property method `job` from `JobOperation` into `_JobOperation`.
We don't have the attribute `_job` in `JobOperation` anymore. Instead we now have `_jobs` hence to maintain backwards compatibility, this method was implemented. 
Please Note: We don't need the property method `job` for `_JobOperation` (as we don't expose `_JobOperation` to users) hence I moved that method to `JobOperation`.

Reply:
👍